### PR TITLE
perf: speed and memory optimisation of the par-geds dsp and hit scripts

### PR DESCRIPTION
**Pull request by ggmarshall**
Stacked on #41 (`input-validation`) so the diff shows only the optimisation changes — retarget to `main` once #41 merges.

Speed/memory pass over the par-geds-dsp calibration scripts (eopt, dplms, evtsel, pz, nopt, svm-build) and the par-geds-hit scripts (qc, ecal). Every change is **bit-identical** — same events selected, same outputs — except (a) the opt-in `use_log_pdf` config knob (default off) and (b) the qc rng seed, both detailed below.

## Shared utilities

- **`get_is_recovering_mask`** replaces the `O(n_discharges × n_events)` discharge-recovery loops in evtsel/pz/qc (which allocated several full-length temporaries per discharge; the qc copies ran on pandas Series over the full calibration run). Exactly loop-equal: a sorted search collects a slightly widened candidate superset of (event, discharge) pairs, then the original floating-point conditions are evaluated on those pairs — tested against a verbatim copy of the old loop, including window-boundary values and the current production call pattern.
- **`take_table_rows`**: in-memory row subset of an lgdo `Table`, row-identical to re-reading the same rows from disk.
- **`get_pulser_mask`**: `np.append` accumulation (quadratic) → single `np.concatenate`, same order/dtype promotion.

## Per script

| Script | Changes |
|---|---|
| evtsel | `read_n_rows` hoisted out of the file×peak loop (was n_peaks× per file — real win on networked filesystems); vectorized recovery mask; the all-events scalar table and the 10k-waveform cut-derivation buffers are freed before the long file×peak loop — the largest peak-RSS cut for the `mem_swap=70` rule; single `count_nonzero` per flush |
| pz | pandas conversion dropped (pure numpy masking); vectorized recovery mask; scalar table freed before the waveform read; second waveform disk read → in-memory subset, **verified bit-identical vs the disk re-read on real p03 cal data (all 39 columns)**, incl. the boolean-mask + `n_rows`-cap semantics of `get_cut_indexes` |
| dplms | second FFT disk read → in-memory subset, **verified bit-identical on real p08 fft data (all 39 columns)**; DSP output and index intermediates freed before the long `dplms_ge_dict` call; dead commented code removed |
| eopt | init `build_dsp` output freed once `dt_eff` is attached (was held for the whole optimisation); `np.where` peak-index lookup → `zip` |
| nopt | per-event energy array freed once baseline indices exist (deliberate read-twice + `del` design kept) |
| svm-build | `field_mask` so only the two training columns load |
| qc | both discharge-recovery loops → `get_is_recovering_mask` (bit-identical); cut-derivation event sample **seeded** — see below |
| ecal | `use_log_pdf` config knob for both `hpge_fit_energy_peaks` calls, `inspect.signature`-guarded so older pygama versions warn and fall back |
| aoe | `use_log_pdf` config knob → `CalAoE` (A/E peak fits + survival-fraction sweeps), same guard |
| lq | `use_log_pdf` config knob → `LQCal` (survival-fraction sweeps; LQ's own fits are binned), same guard |

## `use_log_pdf` (opt-in, default off)

New optional config key for the eopt, dplms, ecal, aoe, and lq configs, forwarded into the pygama fit machinery: with a pygama version containing legend-exp/pygama#704 / #705 / #706, the unbinned NLL fits run in iminuit's `log=True` mode. Measured on real data (full console scripts on the p16 skim fixture unless noted):

- eopt FOM (`fom_fwhm_with_alpha_fit`, ~96% of eopt runtime): **1.71×** faster, `alpha` bit-identical, `fwhm` 5.5×10⁻¹⁰ relative.
- ecal: **1.90×** (15.7 s → 8.3 s), Qbb FWHM shifts 1.2×10⁻⁴ relative (~0.4 eV, far below its uncertainty).
- aoe: **1.45×** (22.0 s → 15.2 s); fitted quantities agree at machine precision, but the derived low-side cut (a rounded grid-crossing of the sigmoid fit) moved one 0.01 step on this low-statistics 2-channel sample, shifting the DEP survival fraction by 0.2 pp — the clearest illustration of why the flag needs physics validation.
- lq: **1.88×** (11.3 s → 6.0 s); results agree to <10⁻⁶ (one `sf_err` at 5×10⁻⁶).

Older pygama versions ignore the key (eopt/dplms) or warn and fall back (ecal/aoe/lq); enabling it in dataprod configs needs physics sign-off.

## qc rng seed (behavior change, one commit)

The 4000-event sample used to derive the qc cut boundaries was drawn with an **unseeded** generator, so repeated qc runs on identical data produced different cut values. It is now seeded (`default_rng(1)`); two runs of the real script on the p16 skim verified byte-identical afterwards. One-time effect: the first seeded production run selects a different (but henceforth reproducible) sample than an unseeded run would have.

## Verification

- New unit tests: `test_discharge.py` (loop-equality incl. edge cases), `test_lgdo_utils.py` (int-index/bool-mask subsets, attrs, unsupported types).
- Real-data bit-identity checks for both in-memory-subset replacements (39/39 columns each) and byte-identity for repeated qc runs.
- Full suite incl. the integration chain (qc/ecal/aoe/lq on the p16 skim) green; pre-commit clean.

Known pre-existing issue deliberately **not** changed here: `evtsel.py` / `pz.py` / `qc.py` compute `discharge_timestamps = np.where(timestamps[discharges])[0]`, which yields positional indices rather than timestamps, so the recovery mask is effectively always all-False. The refactor preserves those semantics bit-for-bit; the one-line fix changes physics selection and should be its own reviewed change. (The same pattern also exists upstream in `pygama.evt.modules.geds` and in legend-dataflow's `pht/qc_phy.py`.)

Disclosure per `AI_POLICY.md`: developed with AI assistance (Claude) and reviewed by the submitter.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/legenddataflowscripts/par/geds/dsp/pz.py [hunk 1]
```diff
@@ -189,11 +185,13 @@ def par_geds_dsp_pz() -> None:
...
-                idx=cuts[: 2 * kwarg_dict["n_events"]][idxs],
-                n_rows=kwarg_dict.pop("n_events"),
+            # tb_data already holds exactly the rows cuts[:2*n_events], so
+            # subset it in memory instead of re-reading the raw files
+            # (row-identical: the read idx was cuts[:2*n_events][idxs] with
+            # the first n_events kept, both in ascending row order)
+            del tb_out
+            tb_data = take_table_rows(
+                tb_data, np.where(idxs)[0][: kwarg_dict.pop("n_events")]
             )
```
[the remark this answers is not in the document]
Comment (ggmarshall) [reply src/legenddataflowscripts/par/geds/dsp/pz.py]: Addressed in 2add74a. For the record: `get_cut_indexes` returns a boolean mask in every supported pygama version (verified back to v2.4.1 — the `idxs[cut_idxs]` in nopt.py is boolean-masking an integer array, which is valid), so the original `np.where(idxs)[0]` was correct. The code now handles both a mask and an index array anyway, since the guard is cheap and future-proofs against an upstream return-type change.

File: src/legenddataflowscripts/par/geds/dsp/eopt.py [hunk 2]
```diff
@@ -144,10 +148,7 @@ def par_geds_dsp_eopt() -> None:
         kwarg_dicts_cusp = []
         kwarg_dicts_trap = []
         kwarg_dicts_zac = []
-        for peak in peaks_kev:
-            peak_idx = np.where(peaks_kev == peak)[0][0]
-            kev_width = kev_widths[peak_idx]
-
+        for peak, kev_width in zip(peaks_kev, kev_widths, strict=False):
```
[the remark this answers is not in the document]
Comment (ggmarshall) [reply src/legenddataflowscripts/par/geds/dsp/eopt.py]: Good catch — switched to `strict=True` in 2add74a. The pre-refactor indexed lookup raised `IndexError` on a short `kev_widths`, so `strict=False` had actually weakened the failure mode; strict restores fail-fast on malformed configs.
